File: src/main/java/net/atos/zac/mail/model/Bronnen.java
package net.atos.zac.mail.model;

import org.flowable.task.api.TaskInfo;

import net.atos.client.zgw.drc.model.generated.EnkelvoudigInformatieObject;
import net.atos.client.zgw.zrc.model.Zaak;

public class Bronnen {

    public final Zaak zaak;

    public final EnkelvoudigInformatieObject document;

    public final TaskInfo taskInfo;

    private Bronnen(final Zaak zaak, final EnkelvoudigInformatieObject document,
            final TaskInfo taskInfo) {
        this.zaak = zaak;
        this.document = document;
        this.taskInfo = taskInfo;
    }

    public static Bronnen fromZaak(final Zaak zaak) {
        return new Builder().add(zaak).build();
    }

    public static Bronnen fromDocument(final EnkelvoudigInformatieObject document) {
        return new Builder().add(document).build();
    }

    public static Bronnen fromTaak(final TaskInfo taskInfo) {
        return new Builder().add(taskInfo).build();
    }

    public static class Builder {
        private Zaak zaak;

        private EnkelvoudigInformatieObject document;

        private TaskInfo taskInfo;

        public Builder add(final Zaak zaak) {
            this.zaak = zaak;
            return this;
        }

        public Builder add(final EnkelvoudigInformatieObject document) {
            this.document = document;
            return this;
        }

        public Builder add(final TaskInfo taskInfo) {
            this.taskInfo = taskInfo;
            return this;
        }

        public Bronnen build() {
            return new Bronnen(zaak, document, taskInfo);
        }
    }
}

File: src/main/kotlin/net/atos/zac/mail/model/Bronnen.kt
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.mail.model

import net.atos.client.zgw.drc.model.generated.EnkelvoudigInformatieObject
import net.atos.client.zgw.zrc.model.Zaak
import org.flowable.task.api.TaskInfo

class Bronnen private constructor(
    val zaak: Zaak?,
    val document: EnkelvoudigInformatieObject?,
    val taskInfo: TaskInfo?
) {
    class Builder {
        private var zaak: Zaak? = null
        private var document: EnkelvoudigInformatieObject? = null
        private var taskInfo: TaskInfo? = null

        fun add(zaak: Zaak) = apply { this.zaak = zaak }
        fun add(document: EnkelvoudigInformatieObject) = apply { this.document = document }
        fun add(taskInfo: TaskInfo) = apply { this.taskInfo = taskInfo }
        fun build() = Bronnen(zaak, document, taskInfo)
    }
}

fun Zaak.getBronnenFromZaak() = Bronnen.Builder().add(this).build()
